File: CPP/ByRound/1993/1993_B_Parity_and_Sum.cpp
// Problem: CF 1993 B - Parity and Sum
// https://codeforces.com/contest/1993/problem/B

/*
B. Parity and Sum
Problem Description:
Given an array of positive integers, we can perform operations where we pick two elements with distinct parities (one odd, one even), and replace the smaller one with their sum. The goal is to find the minimum number of such operations to make all elements have the same parity.

Approach:
- Odd numbers cannot be made even through addition with even numbers.
- Even numbers can be changed into odd numbers by adding them to an odd number.
- Strategy:
  1. Collect all even numbers in a vector.
  2. Track the maximum odd number encountered.
  3. If all numbers are already of same parity, return 0.
  4. Otherwise, simulate merging operations to try reducing the count of distinct parities.

Time Complexity: O(n log n) per test case due to sorting.
Space Complexity: O(n) for storing the even numbers in a vector.

Algorithms/Techniques:
- Greedy approach with sorting
- Simulation of parity-changing operations
*/

#include <algorithm>
#include <cstdio>
#include <vector>
int main() {
  long t;
  scanf("%ld", &t);
  while (t--) {
    long n;
    scanf("%ld", &n);
    std::vector<long> v;           // Vector to store all even numbers
    long mxodd(0);                 // Maximum odd number encountered
    for (long p = 0; p < n; p++) {
      long x;
      scanf("%ld", &x);
      if (x % 2) {                 // If x is odd
        mxodd = (mxodd > x ? mxodd : x);  // Update maximum odd number
      } else {                     // If x is even
        v.push_back(x);            // Add to vector of even numbers
      }
    }
    if (v.size() == n) {           // All elements are odd
      puts("0");
      continue;
    }
    sort(v.begin(), v.end());      // Sort even numbers for greedy merging
    long cnt(v.size());            // Initial count of even numbers
    for (long p = 0; p < v.size(); p++) {
      if (mxodd > v.back()) {      // If max odd is greater than largest even, no gain
        break;
      } else if (mxodd < v[p]) {   // If max odd is smaller than current even, merge and increase count
        ++cnt;
        break;
      } else {                     // Add current even to max odd
        mxodd += v[p];
      }
    }
    printf("%ld\n", cnt);
  }
}


// https://github.com/VaHiX/codeForces/

File: Python/ByRound/1993/1993_B_Parity_and_Sum.py
# Problem: CF 1993 B - Parity and Sum
# https://codeforces.com/contest/1993/problem/B

"""
Algorithm: Greedy approach to minimize operations needed to make all elements of the array have the same parity.
Techniques: 
- Separate even and odd numbers.
- Sort even numbers.
- Greedily merge odd numbers with even numbers based on their values.

Time Complexity: O(n log n) due to sorting the even numbers, where n is the size of the array.
Space Complexity: O(n) for storing even and odd numbers separately.

Approach:
1. Split the input array into even and odd numbers.
2. If there are no even or no odd numbers, no operations are needed.
3. Otherwise, sort the even numbers.
4. Compare the maximum odd number with the smallest and largest even number to decide how many operations are needed:
   - If the max odd is greater than the largest even, we need to add all even numbers to the odd.
   - If the max odd is smaller than the smallest even, we need to add all even numbers + 1 (to handle the odd).
   - Otherwise, simulate adding numbers greedily from even to odd.
"""

class Solution:
    def __init__(self):
        pass

    def solve(self):
        t = int(input())
        res = []
        for _ in range(t):
            int(input())
            arr = list(map(int, input().split()))
            cnt = 0
            even, odd = [], []
            for a in arr:
                if a % 2 == 0:
                    even.append(a)
                else:
                    odd.append(a)
            if len(even) != 0 and len(odd) != 0:
                even = sorted(even)
                odd_max = max(odd)
                if odd_max > even[-1]:  # Max odd is greater than largest even
                    cnt = len(even)
                elif odd_max < even[0]:  # Max odd is smaller than smallest even
                    cnt = len(even) + 1
                else:
                    flag = True
                    for x in even:
                        if odd_max > x:
                            odd_max += x
                        else:
                            flag = False
                            break
                    cnt = len(even)
                    if not flag:
                        cnt += 1
            res.append(cnt)
        return res


if __name__ == "__main__":
    sol = Solution()
    res = sol.solve()
    for x in res:
        print(x)


# https://github.com/VaHiX/CodeForces/